### ISSUE TITLE
Regent: metaprogram task annotations

### ISSUE THREAD

**Issue by syamajala** [StanfordLegion/legion#603]
I would like to be able metaprogram task annotations.
Maybe do something like this:
```
local annotation = rquote end
if variant == "gpu" then
    annotation = rquote __demand(__cuda) end
end

local [annotation] task SomeTask
end
```

**Comment by magnatelee** [comment]
I've been contemplating the syntax for registering task variants, which would support meta-programming better than pragmas. The syntax would look like this:

```
variant SomeTask_cuda of SomeTask where isa(cuda) end
```

And of course the idea here is to specify layout constraints for a variant in `where ... end`. Does this syntax sound reasonable to you, or do you somehow still want to meta-program pragmas?

**Comment by magnatelee** [comment]
Just to clarify, here is the syntax:
```
variant VariantName of TaskName where constraint1, constraint2, ... end
```

**Comment by syamajala** [comment]
This syntax will work for me.

**Comment by elliottslaughter** [comment]
Just to be clear though, we should also allow metaprogramming the annotations on regular tasks since there are use cases for those besides what you get via variants.

**Comment by syamajala** [comment]
I agree with @elliottslaughter I do actually have a use case for this besides just `__demand(__cuda)`.

**Comment by syamajala** [comment]
@magnatelee any word on the new syntax for registering task variants?

**Comment by magnatelee** [comment]
@syamajala no updates yet. I'm planning to add it by the 20.03 milestone. If you want it to be done earlier than that, just let me know.

**Comment by syamajala** [comment]
@magnatelee I will need this sooner rather than later.

**Comment by magnatelee** [comment]
Another useful thing to have with the task variant syntax is a way to allow different task bodies in task variants. So far we assume that a task has only one platform agnostic definition and the compiler automatically generates its variants for different architectures. In scenarios where there exist specialized implementations for certain architectures, such as Singe kernels in S3D, we want to make variants have different implementations.

**Comment by syamajala** [comment]
I think having different task bodies would also be useful for implementing something like the subrank task variant in the Regent version of S3D. 

Just looking at the history of this bug report though, as @elliottslaughter said, the original request of being able to metaprogram task annotations would still be useful. If we could do that then for S3D we could have a build option that optimizes for compilation time vs one that optimizes for application run time performance just by changing the annotations on the tasks. 

So we should probably do both things.

**Comment by elliottslaughter** [comment]
I think we should split this issue into two: one that matches the current title, and one to add an explicit task variant syntax.
